src/lib/auth/redux/authSlice.js

```diff
@@ -0,0 +1,245 @@
+import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
+import { buildUrlEncodedData } from "../utils";
+import { LS_BENTO_WAS_SIGNED_IN, setLSNotSignedIn } from "../performAuth";
+import { decodeJwt } from "jose";
+import { makeResourceKey } from "../resources";
+
+export const tokenHandoff = createAsyncThunk(
+    "auth/TOKEN_HANDOFF",
+    async ({ code, verifier, clientId, authCallbackUrl }, { getState }) => {
+        const url = getState().openIdConfiguration.data?.["token_endpoint"];
+
+        const response = await fetch(url, {
+            method: "POST",
+            headers: { "Content-Type": "application/x-www-form-urlencoded" },
+            body: buildUrlEncodedData({
+                grant_type: "authorization_code",
+                code,
+                client_id: clientId,
+                redirect_uri: authCallbackUrl,
+                code_verifier: verifier,
+            }),
+        });
+
+        // Assuming the server responds with JSON
+        return await response.json();
+    }
+);
+
+export const refreshTokens = createAsyncThunk(
+    "auth/REFRESH_TOKENS",
+    async ({ clientId }, { getState }) => {
+        if (!getState().auth.refreshToken) return;
+
+        const url = getState().openIdConfiguration.data?.["token_endpoint"];
```

Comment:
check if we have `token_endpoint` in condition, and if we don't, return false in the condition + add a console.error saying that we're missing `token_endpoint`/openID config data. then, we don't need `data?.` since we should just error if we don't have the endpoint.